cmd/lifecycle/restorer.go

```diff
@@ -111,64 +142,84 @@ func (r *restoreCmd) Exec() error {
 
 	var appMeta platform.LayersMetadata
 	if r.restoresLayerMetadata() {
-		amd, err := platform.ReadAnalyzed(r.AnalyzedPath, cmd.DefaultLogger)
-		if err == nil {
-			appMeta = amd.Metadata
-		}
+		appMeta = analyzedMD.Metadata
 	}
 
 	return r.restore(appMeta, group, cacheStore)
 }
 
-func (r *restoreCmd) supportsBuildImageExtension() bool {
-	return r.PlatformAPI.AtLeast("0.10")
+func needsPulling(runImage *platform.RunImage) bool {
+	return runImage != nil && runImage.Extend
 }
 
-func (r *restoreCmd) pullBuilderImageIfNeeded() error {
-	if r.BuildImageRef == "" {
-		return nil
+func needsUpdating(runImage *platform.RunImage) bool {
+	if runImage == nil {
+		return false
 	}
-	if _, err := os.Stat(kanikoDir); err != nil {
-		if !os.IsNotExist(err) {
-			return fmt.Errorf("failed to read kaniko directory: %w", err)
-		}
-		return nil
+	if runImage.TargetMetadata == nil {
+		return true
 	}
-	ref, authr, err := auth.ReferenceForRepoName(r.keychain, r.BuildImageRef)
+	digest, err := name.NewDigest(runImage.Reference)
 	if err != nil {
-		return fmt.Errorf("failed to get reference: %w", err)
+		return true
 	}
-	remoteImage, err := remote.Image(ref, remote.WithAuth(authr))
-	if err != nil {
-		return fmt.Errorf("failed to read image: %w", err)
+	return digest.DigestStr() == ""
+}
+
+func (r *restoreCmd) supportsBuildImageExtension() bool {
+	return r.PlatformAPI.AtLeast("0.10")
+}
+
+func (r *restoreCmd) supportsRunImageExtension() bool {
+	return r.PlatformAPI.AtLeast("0.12")
+}
+
+func (r *restoreCmd) pullSparse(imageRef string) (image v1.Image, digest name.Digest, err error) {
+	if imageRef == "" {
+		return nil, name.Digest{}, nil
+	}
+	baseCacheDir := filepath.Join(kanikoDir, "cache", "base")
+	if err := os.MkdirAll(baseCacheDir, 0755); err != nil {
+		return nil, name.Digest{}, fmt.Errorf("failed to create cache directory: %w", err)
 	}
-	buildImageHash, err := remoteImage.Digest()
+	// get remote image
+	remoteImage, digest, err := newRemoteImage(imageRef, r.keychain)
```

Comment:
If exporting to layout format we could look for the run image in the layout directory, but since the pack implementation currently requires the builder to exist in a registry (even when exporting to a daemon) we could keep this requirement and drop it whenever https://github.com/buildpacks/pack/issues/1623 is completed.